File: src/Java/BagQueueStack/Stack.java
package Java.BagQueueStack;

// Simple Pushdown stack (typical stack) implementation.

import java.io.BufferedReader;
import java.util.Iterator;
import java.util.Scanner;

public class Stack<Item> implements Iterable<Item> {
    private int size;
    private Node<Item> current;

    private class Node<Item> {
        private Item data;
        private Node<Item> pointer;
    }

    /**
     * Default no-args constructor for initializing Stack for use.
     */
    public Stack() {
        this.size = 0;
    }

    /**
     * Add an item to Stack.
     * @param data
     */
    public void push(Item data) {
        if (current == null) {
            current = new Node<Item>();
            current.data = data;
        } else {
            Node<Item> temp = current;
            current = new Node<Item>();
            current.data = data;
            current.pointer = temp;
        }
        size++;
    }

    /**
     * Remove and return most recently added item. (Return and remove item from stack based on LIFO policy.)
     * @return item on very top of the current stack.
     */
    public Item pop() {
        if (current == null) throw new NullPointerException();
        Node<Item> temp = current;
        current = current.pointer;
        size--;
        return temp.data;
    }

    /**
     * Examine current stack status. (whether stack is empty or not.)
     * @return {@code true} if stack is empty, {@code false} if stack is not empty.
     */
    public boolean isEmpty() {
        return this.size() != 0;
    }

    /**
     * Return total number of items in Stack.
     * @return current size of stack.
     */
    public int size() {
        return this.size;
    }

    @Override
    public Iterator<Item> iterator() {
        return null;
    }

    // Test Client Code. - Dijkastra's Two-Stack Algorithm for Expression Evaluation.)
    public static void main(String[] args) {
        Stack<String> operator = new Stack<>();
        Stack<Double> operand = new Stack<>();

        Scanner sc = new Scanner(System.in);
        String[] expression = sc.nextLine().split(" ");

        for (String s: expression) {
            if (s.equals("(")); // Ignore Left-parentheses
            else if (s.equals("+")) operator.push(s);
            else if (s.equals("-")) operator.push(s);
            else if (s.equals("*")) operator.push(s);
            else if (s.equals("/")) operator.push(s);
            else if (s.equals("sqrt")) operator.push(s);
            else if (s.equals(")")) {
                double value = operand.pop();
                String currentOperator = operator.pop();
                if (currentOperator.equals("+")) value = operand.pop() + value;
                else if (currentOperator.equals("-")) value = operand.pop() - value;
                else if (currentOperator.equals("*")) value = operand.pop() * value;
                else if (currentOperator.equals("/")) value = operand.pop() / value;
                else if (currentOperator.equals("sqrt")) value = Math.sqrt(value);
                operand.push(value);
            }
            else {
                try {
                    double temp = Double.parseDouble(s);
                    operand.push(temp);
                } catch (Exception e) {
                    System.out.println("String value cannot be converted into Double type: " + s);
                }
            }
        }
        System.out.println("Result: " + operand.pop());
    }
}

File: src/Kotlin/BagQueueStack/Stack.kt
package Kotlin.BagQueueStack

import java.lang.NullPointerException
import java.util.*
import kotlin.NoSuchElementException

// Simple LIFO Stack Implementation

public class Stack<Item>: Iterable<Item> {
    private var size: Int = 0
    private var current: Node<Item>? = null

    private class Node<Item>(var data: Item, var pointer: Node<Item>?)

    /**
     * Add new element to the stack, under the LIFO policy.
     */
    fun push(data: Item): Unit {
        if (current == null) {
            current = Node<Item>(data, null)
            size++
        } else {
            val temp: Node<Item>? = current
            current = Node<Item>(data, null)
            current?.pointer = temp
            size++
        }
    }

    /**
     * Remove and return the most recently added element from stack.
     * @return Most recently added element.
     * @exception NoSuchElementExcepton will be thrown if there is no more element left on stack.
     * @exception NullPointerException will be thrown if returned vale is null.
     */
    fun pop(): Item {
        if (current == null) throw NoSuchElementException()
        val temp: Node<Item>? = current
        current = current?.pointer
        size--
        return temp!!.data
    }

    /**
     * Check stack is currently empty or not.
     * @return true, if stack is currently empty, false, if stack has more value to access.
     */
    fun isEmpty(): Boolean {
        return size() == 0
    }

    /**
     * Return size of current stack.
     * @return size of current stack.
     */
    fun size(): Int {
        return size
    }

    override fun iterator(): Iterator<Item> {

        class LinkedStackIterator(var node: Node<Item>?): Iterator<Item> {

            /**
             * Check whether iterator has more value to access.
             * @return ture, if iterator has next value to access, false, if iterator has no more value to access.
             */
            override fun hasNext(): Boolean {
                return node?.pointer != null
            }

            /**
             * Return next value in iterator.
             * @return next element in this iterator.
             * @exception NullPointerException will be thrown, if there are no more elements to access, or, if return
             * value has a null value.
             */
            override fun next(): Item {
                if (!hasNext()) throw NullPointerException()
                val temp: Node<Item>? = node
                node = node?.pointer
                return temp!!.data
            }
        }
        return LinkedStackIterator(current)
    }

}

// Client Code for Unit Test.
fun main(args: Array<String>) {
    // dijkstraTwoStackAlgorithm(args)
    stackOrderTest(args)
}

fun dijkstraTwoStackAlgorithm(args: Array<String>) {
    val operator: Stack<String> = Stack<String>()
    val operand: Stack<Double> = Stack<Double>()
    val sc = Scanner(System.`in`)
    val inputString: List<String> = sc.nextLine().split(" ")
    sc.close()

    for (s: String in inputString) {
        if (s.equals("("));
        else if (s.equals("+")) operator.push(s)
        else if (s.equals("-")) operator.push(s)
        else if (s.equals("/")) operator.push(s)
        else if (s.equals("*")) operator.push(s)
        else if (s.equals("sqrt")) operator.push(s)
        else if (s.equals(")")) {
            var num = operand.pop()
            val poppedOperator = operator.pop()
            if (poppedOperator.equals("+")) num += operand.pop()
            else if (poppedOperator.equals("-")) num -= operand.pop()
            else if (poppedOperator.equals("/")) num /= operand.pop()
            else if (poppedOperator.equals("*")) num *= operand.pop()
            operand.push(num)
        }
        else {
            try {
                operand.push(s.toDouble())
            } catch (ime: InputMismatchException) {
                println(ime.stackTrace)
            } catch (e: Exception) {
                println(e.stackTrace)
            }
        }
    }
    println("Result: ${operand.pop()}")
}

fun stackOrderTest(args: Array<String>) {
    val stack: Stack<String> = Stack<String>()
    val sc = Scanner(System.`in`)
    var inputString: Iterator<String> = sc.nextLine().split(" ").iterator()
    while (inputString.hasNext()) {
        var s = inputString.next()
        if (!s.equals("-")) {
            stack.push(s)
        } else if (!stack.isEmpty()) {
            print(stack.pop() + " ")
        }
    }

    println("(${stack.size()} left on stack.")
}